frontends/p4/unusedDeclarations.cpp

```diff
@@ -60,138 +62,162 @@ void UsedDeclSet::dbprint(std::ostream &out) const {
     auto orig = getOriginal<IR::P4Parser>();
     if (!used.isUsed(orig)) {
         if (giveWarning(orig))
-            warn(ErrorType::WARN_UNUSED, "Parser %2% is not used; removing", parser,
-                 parser->externalName());
-        LOG3("Removing " << parser << dbp(orig));
-        prune();
-        return nullptr;
+            warn(ErrorType::WARN_UNUSED, "parser '%2%' is unused", parser, parser->externalName());
+        if (removeUnused) {
+            if (warnRemoved) {
+                info(ErrorType::INFO_REMOVED, "removing parser '%2%'", parser,
+                     parser->externalName());
+            }
+            prune();
+            return nullptr;
+        }
     }
-
     visit(parser->parserLocals, "parserLocals");
     visit(parser->states, "states");
     prune();
     return parser;
 }
 
-const IR::Node *RemoveUnusedDeclarations::preorder(IR::P4Table *table) {
+const IR::Node *UnusedDeclarations::preorder(IR::P4Table *table) {
     if (!used.isUsed(getOriginal<IR::IDeclaration>())) {
         if (giveWarning(getOriginal()))
-            warn(ErrorType::WARN_UNUSED, "Table %1% is not used; removing", table);
-        LOG3("Removing " << table);
-        table = nullptr;
+            warn(ErrorType::WARN_UNUSED, "table '%1%' is unused", table);
+        if (removeUnused) {
+            if (warnRemoved) {
+                info(ErrorType::INFO_REMOVED, "removing table '%2%'", table, table->externalName());
+            }
+            prune();
+            return nullptr;
+        }
     }
     prune();
     return table;
 }
 
-const IR::Node *RemoveUnusedDeclarations::preorder(IR::Declaration_Variable *decl) {
+const IR::Node *UnusedDeclarations::preorder(IR::Declaration_Variable *decl) {
     prune();
+    if (!removeUnused) return process(decl);
     if (decl->initializer == nullptr) return process(decl);
     if (!SideEffects::check(decl->initializer, this, nullptr, nullptr)) return process(decl);
     return decl;
 }
 
-const IR::Node *RemoveUnusedDeclarations::process(const IR::IDeclaration *decl) {
+const IR::Node *UnusedDeclarations::process(const IR::IDeclaration *decl) {
     LOG3("Visiting " << decl);
     if (decl->getName().name == IR::ParserState::verify && getParent<IR::P4Program>())
         return decl->getNode();
-    if (decl->externalName(decl->getName().name).startsWith("__"))
+    // @name annotations can be unparsed until the first ConstantFolding. In case
+    // of UnusedDeclarations in the scope of WarnAboutUnusedDeclarations, we cannot
+    // use externalName as it expects the @name annotation to be a single string.
+    bool skipNameCheck = false;
+    if (const auto *annotated = decl->to<IR::IAnnotated>()) {
+        if (const auto *anno = annotated->getAnnotation(IR::Annotation::nameAnnotation)) {
+            skipNameCheck = anno->getSingleString(false) == cstring::empty;
+        }
+    }
+    if (!skipNameCheck && decl->externalName(decl->getName().name).startsWith("__"))
```

Comment:
Maybe we should fix `externalName` to work with unparsed annotations.

Reply:
You could have `externalName` return `cstring()` if the annotation exists and is not parsed (as it does currently if it *is* parsed, but the expression is not a StringLiteral).  Both of these are a bit of a hack -- returning a null cstring as a "don't know" indication.